jira/client.py

```diff
@@ -2223,59 +2223,23 @@ def _get_user_identifier(self, user: User) -> str:
         """
         return user.accountId if self._is_cloud else user.name
 
-    def _get_user_id(self, user: str | None) -> str | None:
-        """Internal method for translating a user search (str) to an id.
-
-        Return None and -1 unchanged.
-
-        This function uses :py:meth:`JIRA.search_users` to find the user and then using :py:meth:`JIRA._get_user_identifier` extracts
-        the relevant identifier property depending on whether the instance is a Cloud or self-hosted Instance.
-
-        Args:
-            user (Optional[str]): The search term used for finding a user. None, '-1' and -1 are equivalent to 'Unassigned'.
-
-        Raises:
-            JIRAError: If any error occurs.
-
-        Returns:
-            Optional[str]: The Jira user's identifier. Or "-1" and None unchanged.
-        """
-        if user in (None, -1, "-1"):
-            return user
-        try:
-            user_obj: User
-            if self._is_cloud:
-                users = self.search_users(query=user, maxResults=20)
-            else:
-                users = self.search_users(user=user, maxResults=20)
-
-            if len(users) < 1:
-                raise JIRAError(f"No matching user found for: '{user}'")
-
-            matches = []
-            if len(users) > 1:
-                matches = [u for u in users if self._get_user_identifier(u) == user]
-            user_obj = matches[0] if matches else users[0]
-
-        except Exception as e:
-            raise JIRAError(str(e))
-        return self._get_user_identifier(user_obj)
-
     # non-resource
     @translate_resource_args
-    def assign_issue(self, issue: int | str, assignee: str | None) -> bool:
+    def assign_issue(self, issue: int | str, assignee: str | None | User) -> bool:
         """Assign an issue to a user.
 
         Args:
             issue (Union[int, str]): the issue ID or key to assign
-            assignee (str): the user to assign the issue to. None will set it to unassigned. -1 will set it to Automatic.
+            assignee (Union[str, User]): username (for hosted) or account ID (for cloud) of the user to add to the
+                watchers list. Alternatively, you can provide the User object itself.
 
         Returns:
             bool
         """
         url = self._get_latest_url(f"issue/{issue}/assignee")
-        user_id = self._get_user_id(assignee)
-        payload = {"accountId": user_id} if self._is_cloud else {"name": user_id}
+        if isinstance(assignee, User):
+            assignee = self.get_user_identifier(assignee)
+        payload = {"accountId": assignee} if self._is_cloud else {"name": assignee}
```

Comment:
```suggestion
        if isinstance(assignee, User):
            user_id = self.get_user_identifier(assignee)
        payload = {"accountId": user_id} if self._is_cloud else {"name": user_id}
```

(non-blocking) I generally have a preference to avoid overriding function arguments values inside the function. The above is an example of what I would prefer

Reply:
also applies to `watcher` below

Reply:
I'm not opposed to that idea, the only reason why I opted for value overriding is because `assignee` might also be a string or None and in that case it's being passed directly into the payload dict as-is. I think in order to maintain the value of the variable untouched, the code would have to change to something like:
```suggestion
        if isinstance(assignee, User):
            user_id = self.get_user_identifier(assignee)
        else:
            user_id = assignee
        payload = {"accountId": user_id} if self._is_cloud else {"name": user_id}
```
If that's okay with you, I'll go ahead and change it, I just figured the current design might be a bit tidier, but I'm not dead set on it.

Reply:
this is a good point, let's leave it as is for now